File: PythonNBA/scripts/backfill_last_n_days.py
#!/usr/bin/env python3
# scripts/backfill_last_n_days.py

import json
import math
import os
import sys
import time
from datetime import datetime, timedelta

from sources.nba_stats import fetch_nba_stats, fetch_nba_stats_enhanced
from sources.blend_stats import blend_base, blend_for_game
from sources.espn_scoreboard import fetch_scoreboard, extract_final_scores
from sources.teamrankings_trends import fetch_ats_trends, fetch_ou_trends
from sources.rest_detect import apply_b2b_adjustment

from model_engine import load_defaults, get_avgs, analyze_game
from store import load_store, save_store, upsert_run
from self_tune import tune_weights, compute_residual_var
from kalman_state import (
    load_kalman_state, save_kalman_state, initialize_kalman,
    apply_daily_drift, batch_update, prune_processed_games,
)

from sources.odds_batch_historical import fetch_odds_for_day

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
CACHE_DIR = os.path.join(SCRIPT_DIR, "..", "data", "stats_cache")

# Days at the START of the window (oldest) used for warm-up only -- picks not evaluated
BURN_IN_DAYS = 20

# Minimum games a team must have played before we trust its stats
MIN_GAMES = 25


def sleep(seconds):
    time.sleep(seconds)


def recency_weight(days_ago):
    if days_ago <= 15:
        return 1.0
    if days_ago <= 30:
        return 0.75
    if days_ago <= 45:
        return 0.5
    return 0.25


def to_display_date(yyyymmdd_str):
    return f"{yyyymmdd_str[:4]}-{yyyymmdd_str[4:6]}-{yyyymmdd_str[6:8]}"


def yyyymmdd_from_date(d):
    return f"{d.year:04d}{d.month:02d}{d.day:02d}"


def date_minus_days_central(days_ago):
    import pytz
    tz = pytz.timezone("America/Chicago")
    now = datetime.now(tz)
    base = datetime(now.year, now.month, now.day) - timedelta(days=days_ago)
    return yyyymmdd_from_date(base)


def pick_home_away_from_scoreboard_event(ev):
    comp = (ev or {}).get("competitions", [{}])[0] if ev else {}
    competitors = comp.get("competitors", [])
    if not isinstance(competitors, list) or len(competitors) != 2:
        return None

    home = None
    away = None
    for c in competitors:
        if c.get("homeAway") == "home":
            home = c
        elif c.get("homeAway") == "away":
            away = c
    if not home or not away:
        return None

    home_team = home.get("team", {})
    away_team = away.get("team", {})
    home_name = str(home_team.get("displayName") or home_team.get("name") or home_team.get("location") or "").strip()
    away_name = str(away_team.get("displayName") or away_team.get("name") or away_team.get("location") or "").strip()

    commence_time_iso = comp.get("date")
    return {"home": home_name, "away": away_name, "commenceTimeIso": commence_time_iso}


def main():
    days = int(sys.argv[1]) if len(sys.argv) > 1 else 60
    if days <= 0:
        print("Usage: python scripts/backfill_last_n_days.py 8")
        sys.exit(1)

    store = load_store()
    defaults = load_defaults()

    ats = fetch_ats_trends()
    ou = fetch_ou_trends()

    # Cache stats per date
    stats_cache = {}

    def get_stats_for_date(date_yyyymmdd):
        if date_yyyymmdd in stats_cache:
            return stats_cache[date_yyyymmdd]

        # Check disk cache first
        os.makedirs(CACHE_DIR, exist_ok=True)
        disk_path = os.path.join(CACHE_DIR, date_yyyymmdd + ".json")
        if os.path.exists(disk_path):
            with open(disk_path, "r") as f:
                raw = json.load(f)
            if "season" in raw and raw.get("last10"):
                stats_cache[date_yyyymmdd] = raw
                return raw
            print(f"  [backfill] Old-format cache for {date_yyyymmdd} -- re-fetching enhanced stats...")

        # Fetch from NBA.com
        date_to = f"{date_yyyymmdd[:4]}-{date_yyyymmdd[4:6]}-{date_yyyymmdd[6:8]}"
        try:
            enhanced = fetch_nba_stats_enhanced(date_to)
        except Exception as e:
            print(f"  [backfill] Enhanced fetch failed: {e} -- trying season only")
            stats_obj = fetch_nba_stats(date_to)
            enhanced = {"season": stats_obj, "last10": None, "home": None, "away": None}
        stats_cache[date_yyyymmdd] = enhanced

        # Write to disk cache
        try:
            with open(disk_path, "w") as f:
                json.dump(enhanced, f)
        except Exception as e:
            print(f"  [backfill] cache write failed: {e}")

        return enhanced

    # Kalman state -- initialized on first date with stats, evolves forward
    kalman_state = None
    prev_date = None
    prev_graded = []       # games with projections -> Kalman + self-tune
    b2b_teams = set()      # teams that played yesterday -> B2B penalty

    for i in range(days, 0, -1):  # stop at yesterday -- never backfill today
        date = date_minus_days_central(i)
        date_display = to_display_date(date)

        base_w = store.get("weights", {})
        if not base_w or not isinstance(base_w, dict) or len(base_w) == 0:
            base_w = defaults["DEFAULT_W"]
        base_w = {**base_w}

        base_w_var = store.get("weightsVar", {})
        if not base_w_var or not isinstance(base_w_var, dict) or len(base_w_var) == 0:
            base_w_var = defaults["DEFAULT_W_VAR"]
        base_w_var = {**base_w_var}

        # Fetch enhanced stats
        try:
            enhanced = get_stats_for_date(date)
            base_stats = blend_base(enhanced["season"], enhanced.get("last10"), base_w.get("recentWeight", 0.35))
            a = get_avgs(base_stats)
        except Exception as e:
            print(f"  [backfill] Could not fetch stats for {date_display}: {e} -- skipping")
            continue

        # Initialize Kalman on first date with stats
        if kalman_state is None:
            kalman_state = initialize_kalman(enhanced["season"])
            kalman_state["lastDriftDate"] = date
            print(f"  [backfill] Kalman initialized from {date_display}")

        # -- Start of new day: process yesterday's results before today's analysis --
        # Flow: Kalman update -> self-tune -> drift -> residualVar -> analyze
        if date != prev_date:
            prev_date = date

            # 1-2. Kalman + self-tune (requires projections -- only completed games)
            if prev_graded:
                batch_update(kalman_state, prev_graded)

                in_burn_in_prev = (i + 1) > (days - BURN_IN_DAYS)
                if not in_burn_in_prev:
                    result = tune_weights(base_w, base_w_var, prev_graded)
                    base_w = result["W"]
                    base_w_var = result["W_var"]
                    store["weights"] = result["W"]
                    store["weightsVar"] = result["W_var"]

            # Build B2B set for TODAY -- teams that played yesterday
            b2b_teams = set()
            for g in prev_graded:
                if g.get("home"):
                    b2b_teams.add(g["home"])
                if g.get("away"):
                    b2b_teams.add(g["away"])

            prev_graded = []

            # Drift AFTER learning -- account for time passing to today
            apply_daily_drift(kalman_state, date)

        # Compute dynamic residualVar from history so far
        dynamic_residual_var = compute_residual_var(store.get("runs", []))

        sb = fetch_scoreboard(date)
        finals = extract_final_scores(sb)
        events = sb.get("events", []) if isinstance(sb, dict) else []

        # Collect all games for batch odds fetch
        games_list = []
        for ev in events:
            ha = pick_home_away_from_scoreboard_event(ev)
            if not ha:
                continue
            f = next((x for x in finals if x["away"] == ha["away"] and x["home"] == ha["home"]), None)
            if not f:
                continue
            games_list.append({**ha, "awayScore": f["awayScore"], "homeScore": f["homeScore"]})

        # Batch fetch ALL odds for this date in 1-2 API calls (cached to disk)
        day_odds = {}
        if games_list:
            try:
                day_odds = fetch_odds_for_day(date, games_list)
            except Exception as e:
                print(f"  [backfill] Batch odds fetch failed: {e}")

        games = []

        # Apply B2B rest penalty to today's stats (derived from yesterday's games)
        result = apply_b2b_adjustment(base_stats, b2b_teams, games_list)
        adjusted_stats = result["adjusted"]
        b2b_notes = result["b2bNotes"]

        for gl in games_list:
            key = f"{gl['away']}@{gl['home']}"
            odds = day_odds.get(key, {"line": None, "total": None, "_book": None, "_note": "No batch odds"})

            g = {
                "away": gl["away"],
                "home": gl["home"],
                "line": odds.get("line"),
                "total": odds.get("total"),
                "_book": odds.get("_book"),
            }

            if not isinstance(g["line"], (int, float)) or not isinstance(g["total"], (int, float)):
                games.append({
                    **g,
                    "awayScore": gl["awayScore"],
                    "homeScore": gl["homeScore"],
                    "status": "MISSING_ODDS",
                    "note": odds.get("_note", "Historical odds not available for this game"),
                })
                continue

            game_stats = blend_for_game(
                adjusted_stats, enhanced.get("home"), enhanced.get("away"),
                g["home"], g["away"], base_w.get("locationWeight", 0.25)
            )
            game_avgs = get_avgs(game_stats)

            # No H2H matchups for NBA model -- pass None as extra arg
            r = analyze_game(g, game_stats, game_avgs, base_w, None, kalman_state, base_w_var, dynamic_residual_var, None)
            if not r:
                games.append({
                    **g,
                    "awayScore": gl["awayScore"],
                    "homeScore": gl["homeScore"],
                    "status": "SKIPPED",
                    "note": "analyzeGame returned null (team name mismatch or bad inputs)",
                })
                continue

            r["awayScore"] = gl["awayScore"]
            r["homeScore"] = gl["homeScore"]
            r["_recencyWeight"] = recency_weight(i)

            # B2B notes for display
            away_b2b = b2b_notes.get(g["away"])
            home_b2b = b2b_notes.get(g["home"])
            if away_b2b or home_b2b:
                parts = []
                if away_b2b:
                    parts.append(f"{g['away']}: {away_b2b}")
                if home_b2b:
                    parts.append(f"{g['home']}: {home_b2b}")
                r["b2bNote"] = " | ".join(parts)

            r["trends"] = {
                "away": {
                    "atsPct": (ats.get(r["away"]) or {}).get("atsPct"),
                    "atsPlusMinus": (ats.get(r["away"]) or {}).get("atsPlusMinus"),
                    "overPct": (ou.get(r["away"]) or {}).get("overPct"),
                    "underPct": (ou.get(r["away"]) or {}).get("underPct"),
                    "totalPlusMinus": (ou.get(r["away"]) or {}).get("totalPlusMinus"),
                },
                "home": {
                    "atsPct": (ats.get(r["home"]) or {}).get("atsPct"),
                    "atsPlusMinus": (ats.get(r["home"]) or {}).get("atsPlusMinus"),
                    "overPct": (ou.get(r["home"]) or {}).get("overPct"),
                    "underPct": (ou.get(r["home"]) or {}).get("underPct"),
                    "totalPlusMinus": (ou.get(r["home"]) or {}).get("totalPlusMinus"),
                },
            }

            games.append(r)

        completed = [
            x for x in games
            if x.get("status") not in ("MISSING_ODDS", "SKIPPED")
            and isinstance(x.get("homeScore"), (int, float)) and math.isfinite(x["homeScore"])
            and isinstance(x.get("awayScore"), (int, float)) and math.isfinite(x["awayScore"])
        ]

        in_burn_in = i > (days - BURN_IN_DAYS)

        # Collect graded games for next day's processing
        for g in completed:
            g["_kalmanDate"] = date
        prev_graded = completed

        run = {
            "date": date,
            "dateDisplay": date_display,
            "burnIn": in_burn_in,
            "weightsUsed": {**base_w},
            "weightsNext": {**base_w},
            "games": games,
            "summaryText": "",
        }

        upsert_run(store, run)
        save_store(store)

        # Save Kalman state every day
        if kalman_state:
            save_kalman_state(kalman_state)

        counts = {}
        for x in games:
            k = x.get("status", "OK")
            counts[k] = counts.get(k, 0) + 1
        burn_in_tag = " [BURN-IN]" if in_burn_in else ""
        b2b_count = len(b2b_notes)
        print(
            f"Backfilled {date_display}{burn_in_tag}: games={len(games)}, completed={len(completed)} statuses={json.dumps(counts)}"
            + (f" b2b={b2b_count}" if b2b_count else "")
        )

    # Save final Kalman state
    if kalman_state:
        prune_processed_games(kalman_state, 60)
        save_kalman_state(kalman_state)
        print("Kalman state saved.")

    print("Backfill complete.")


if __name__ == "__main__":
    try:
        main()
    except Exception as err:
        import traceback
        traceback.print_exc()
        sys.exit(1)

File: NBAFullseason/scripts/backfill_last_n_days.mjs
// scripts/backfill_last_n_days.mjs
import { fetchNBAStats, fetchNBAStatsEnhanced } from "./sources/nba_stats.mjs";
import { blendBase, blendForGame } from "./sources/blend_stats.mjs";
import { fetchScoreboard, extractFinalScores } from "./sources/espn_scoreboard.mjs";
import { fetchATSTrends, fetchOUTRends } from "./sources/teamrankings_trends.mjs";
import { applyB2BAdjustment } from "./sources/rest_detect.mjs";

import { loadDefaults, getAvgs, analyzeGame } from "./model_engine.mjs";
import { loadStore, saveStore, upsertRun } from "./store.mjs";
import { tuneWeights, computeResidualVar } from "./self_tune.mjs";
import {
  loadKalmanState, saveKalmanState, initializeKalman,
  applyDailyDrift, batchUpdate, pruneProcessedGames,
} from "./kalman_state.mjs";

import { fetchOddsForDay } from "./sources/odds_batch_historical.mjs";

import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CACHE_DIR = path.join(__dirname, "..", "data", "stats_cache");

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

// Days at the START of the window (oldest) used for warm-up only — picks not evaluated
const BURN_IN_DAYS = 0;  // MIN_GP=15 handles warm-up — count every pick that fires

// Weighted recency: recent games matter more than older ones
function recencyWeight(daysAgo) {
  if (daysAgo <= 15) return 1.0;
  if (daysAgo <= 30) return 0.75;
  if (daysAgo <= 45) return 0.5;
  return 0.25;
}

// Minimum games a team must have played before we trust its stats
const MIN_GAMES = 25;

function toDisplayDate(yyyymmddStr) {
  return `${yyyymmddStr.slice(0, 4)}-${yyyymmddStr.slice(4, 6)}-${yyyymmddStr.slice(6, 8)}`;
}

function yyyymmddFromDate(d) {
  const y = d.getFullYear();
  const m = String(d.getMonth() + 1).padStart(2, "0");
  const day = String(d.getDate()).padStart(2, "0");
  return `${y}${m}${day}`;
}

function dateMinusDaysCentral(daysAgo) {
  const now = new Date();
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/Chicago",
    year: "numeric",
    month: "2-digit",
    day: "2-digit"
  });
  const parts = fmt.formatToParts(now);
  const y = Number(parts.find((p) => p.type === "year").value);
  const m = Number(parts.find((p) => p.type === "month").value);
  const d = Number(parts.find((p) => p.type === "day").value);

  const base = new Date(Date.UTC(y, m - 1, d));
  base.setUTCDate(base.getUTCDate() - daysAgo);
  return yyyymmddFromDate(base);
}

function pickHomeAwayFromScoreboardEvent(ev) {
  const comp = ev?.competitions?.[0];
  const competitors = comp?.competitors;
  if (!Array.isArray(competitors) || competitors.length !== 2) return null;

  const home = competitors.find((c) => c?.homeAway === "home");
  const away = competitors.find((c) => c?.homeAway === "away");
  if (!home || !away) return null;

  const homeName = String(
    home?.team?.displayName || home?.team?.name || home?.team?.location || ""
  ).trim();
  const awayName = String(
    away?.team?.displayName || away?.team?.name || away?.team?.location || ""
  ).trim();

  const commenceTimeIso = comp?.date; // ISO in UTC from ESPN
  return { home: homeName, away: awayName, commenceTimeIso };
}

async function main() {
  const days = Number(process.argv[2] || 60);
  if (!Number.isFinite(days) || days <= 0) {
    console.log("Usage: node scripts/backfill_last_n_days.mjs 8");
    process.exit(1);
  }

  const store = loadStore();
  const defaults = loadDefaults();

  const [ats, ou] = await Promise.all([
    fetchATSTrends(),
    fetchOUTRends()
  ]);

  // Cache stats per date — NBA.com has rate limits so we reuse within same date
  const statsCache = new Map();

  async function getStatsForDate(dateYYYYMMDD) {
    if (statsCache.has(dateYYYYMMDD)) return statsCache.get(dateYYYYMMDD);

    // Check disk cache first (shared with recalculate.mjs)
    if (!fs.existsSync(CACHE_DIR)) fs.mkdirSync(CACHE_DIR, { recursive: true });
    const diskPath = path.join(CACHE_DIR, dateYYYYMMDD + ".json");
    if (fs.existsSync(diskPath)) {
      const raw = JSON.parse(fs.readFileSync(diskPath, "utf8"));
      // If enhanced format with last10 data, use it directly
      if (raw.season && raw.last10) {
        statsCache.set(dateYYYYMMDD, raw);
        return raw;
      }
      // Old format (no last10/home/away) — re-fetch enhanced from NBA.com
      console.log(`  [backfill] Old-format cache for ${dateYYYYMMDD} — re-fetching enhanced stats...`);
    }

    // Fetch enhanced from NBA.com (date-accurate stats as-of that day)
    const dateTo = `${dateYYYYMMDD.slice(0,4)}-${dateYYYYMMDD.slice(4,6)}-${dateYYYYMMDD.slice(6,8)}`;
    let enhanced;
    try {
      enhanced = await fetchNBAStatsEnhanced(dateTo);
    } catch (e) {
      console.warn(`  [backfill] Enhanced fetch failed: ${e.message} — trying season only`);
      const stats = await fetchNBAStats(dateTo);
      enhanced = { season: stats, last10: null, home: null, away: null };
    }
    statsCache.set(dateYYYYMMDD, enhanced);

    // Write to disk cache for recalculate.mjs to reuse
    try { fs.writeFileSync(diskPath, JSON.stringify(enhanced)); }
    catch (e) { console.warn("  [backfill] cache write failed:", e.message); }

    return enhanced;
  }

  // H2H matchup history — builds progressively as days are backfilled.
  // Each date only sees games that happened before it.
  const h2hMatchups = {};

  // Kalman state — initialized on first date with stats, evolves forward
  let kalmanState = null;
  let prevDate = null;
  let prevGraded = [];     // games with projections → Kalman + self-tune
  let prevAllScored = [];  // all games with scores → H2H (includes SKIPPED)
  let b2bTeams = new Set(); // teams that played yesterday → B2B penalty in analyzeGame

  for (let i = days; i >= 0; i--) {
    const date = dateMinusDaysCentral(i);
    const dateDisplay = toDisplayDate(date);

    let baseW    = (store.weights    && Object.keys(store.weights).length    > 0) ? store.weights    : defaults.DEFAULT_W;
    let baseWVar = (store.weightsVar && Object.keys(store.weightsVar).length > 0) ? store.weightsVar : defaults.DEFAULT_W_VAR;

    // Fetch enhanced stats as they were on this specific date
    let enhanced, baseStats, a;
    try {
      enhanced = await getStatsForDate(date);
      baseStats = blendBase(enhanced.season, enhanced.last10, baseW.recentWeight ?? 0.35);
      a = getAvgs(baseStats);
    } catch (e) {
      console.warn(`  [backfill] Could not fetch stats for ${dateDisplay}: ${e.message} — skipping`);
      continue;
    }

    // Initialize Kalman on first date with stats
    if (!kalmanState) {
      kalmanState = initializeKalman(enhanced.season);
      kalmanState.lastDriftDate = date;
      console.log(`  [backfill] Kalman initialized from ${dateDisplay}`);
    }

    // ── Start of new day: process yesterday's results before today's analysis ──
    // Flow: Kalman update → self-tune → H2H → drift → residualVar → analyze
    if (date !== prevDate) {
      prevDate = date;

      // 1-2. Kalman + self-tune (requires projections — only completed games)
      if (prevGraded.length) {
        batchUpdate(kalmanState, prevGraded);

        const { W: tunedW, W_var: tunedWVar } = tuneWeights(baseW, baseWVar, prevGraded);
        baseW = tunedW;
        baseWVar = tunedWVar;
        store.weights = tunedW;
        store.weightsVar = tunedWVar;
      }

      // 3. H2H matchup history (just needs scores — includes early season SKIPPED games)
      for (const g of prevAllScored) {
        const key = [g.home, g.away].sort().join("::");
        if (!h2hMatchups[key]) {
          h2hMatchups[key] = { team1: [g.home, g.away].sort()[0], team2: [g.home, g.away].sort()[1], games: [] };
        }
        h2hMatchups[key].games.push({
          date: g._kalmanDate || date,
          home: { team: g.home, pts: g.homeScore },
          away: { team: g.away, pts: g.awayScore },
        });
      }

      // 5. Build B2B set for TODAY — teams that played yesterday
      b2bTeams = new Set();
      for (const g of prevAllScored) {
        if (g.home) b2bTeams.add(g.home);
        if (g.away) b2bTeams.add(g.away);
      }

      prevGraded = [];
      prevAllScored = [];

      // 4. Drift AFTER learning — account for time passing to today
      applyDailyDrift(kalmanState, date);
    }

    // 4. Compute dynamic residualVar from history so far
    const dynamicResidualVar = computeResidualVar(store.runs || []);

    const sb = await fetchScoreboard(date);
    const finals = extractFinalScores(sb);
    const events = Array.isArray(sb?.events) ? sb.events : [];

    // Collect all games for batch odds fetch
    const gamesList = [];
    for (const ev of events) {
      const ha = pickHomeAwayFromScoreboardEvent(ev);
      if (!ha) continue;
      const f = finals.find((x) => x.away === ha.away && x.home === ha.home);
      if (!f) continue;
      gamesList.push({ ...ha, awayScore: f.awayScore, homeScore: f.homeScore });
    }

    // Batch fetch ALL odds for this date in 1-2 API calls (cached to disk)
    let dayOdds = {};
    if (gamesList.length > 0) {
      try {
        dayOdds = await fetchOddsForDay(date, gamesList);
      } catch (e) {
        console.warn(`  [backfill] Batch odds fetch failed: ${e.message}`);
      }
    }

    const games = [];

    // Apply B2B rest penalty to today's stats (derived from yesterday's games)
    const { adjusted: adjustedStats, b2bNotes } = applyB2BAdjustment(baseStats, b2bTeams, gamesList);

    // Compute per-team deltas for recalculate replay
    const STAT_KEYS = ["OFF", "DEF", "TS", "TO", "ORR", "PACE"];
    function computeTeamDelta(team) {
      const base = baseStats[team];
      const adj  = adjustedStats[team];
      if (!base || !adj || base === adj) return null;
      const d = {};
      let any = false;
      for (const k of STAT_KEYS) {
        const diff = (adj[k] || 0) - (base[k] || 0);
        if (Math.abs(diff) > 0.0001) { d[k] = Math.round(diff * 10000) / 10000; any = true; }
      }
      return any ? d : null;
    }

    for (const gl of gamesList) {
      const key = `${gl.away}@${gl.home}`;
      const odds = dayOdds[key] || { line: null, total: null, _book: null, _note: "No batch odds" };

      const g = {
        away: gl.away,
        home: gl.home,
        line: odds.line,
        total: odds.total,
        _book: odds._book
      };

      if (typeof g.line !== "number" || typeof g.total !== "number") {
        games.push({
          ...g,
          awayScore: gl.awayScore,
          homeScore: gl.homeScore,
          status: "MISSING_ODDS",
          note: odds._note || "Historical odds not available for this game"
        });
        continue;
      }

      const gameStats = blendForGame(
        adjustedStats, enhanced.home, enhanced.away,
        g.home, g.away, baseW.locationWeight ?? 0.25
      );
      const gameAvgs = getAvgs(gameStats);

      const r = analyzeGame(g, gameStats, gameAvgs, baseW, null, kalmanState, baseWVar, dynamicResidualVar, h2hMatchups);
      if (!r) {
        games.push({
          ...g,
          awayScore: gl.awayScore,
          homeScore: gl.homeScore,
          status: "SKIPPED",
          note: "analyzeGame returned null (team name mismatch or bad inputs)"
        });
        continue;
      }

      r.awayScore = gl.awayScore;
      r.homeScore = gl.homeScore;
      r._recencyWeight = recencyWeight(i);

      // Cache B2B deltas for recalculate replay (same pattern as run_daily)
      const awayDelta = computeTeamDelta(g.away);
      const homeDelta = computeTeamDelta(g.home);
      if (awayDelta || homeDelta) {
        r._adjDeltas = {};
        if (awayDelta) r._adjDeltas.away = awayDelta;
        if (homeDelta) r._adjDeltas.home = homeDelta;
      }

      // B2B notes for display
      const awayB2B = b2bNotes[g.away] || null;
      const homeB2B = b2bNotes[g.home] || null;
      if (awayB2B || homeB2B) {
        r.b2bNote = [awayB2B ? `${g.away}: ${awayB2B}` : null, homeB2B ? `${g.home}: ${homeB2B}` : null].filter(Boolean).join(" | ");
      }

      r.trends = {
        away: {
          atsPct: ats?.[r.away]?.atsPct ?? null,
          atsPlusMinus: ats?.[r.away]?.atsPlusMinus ?? null,
          overPct: ou?.[r.away]?.overPct ?? null,
          underPct: ou?.[r.away]?.underPct ?? null,
          totalPlusMinus: ou?.[r.away]?.totalPlusMinus ?? null
        },
        home: {
          atsPct: ats?.[r.home]?.atsPct ?? null,
          atsPlusMinus: ats?.[r.home]?.atsPlusMinus ?? null,
          overPct: ou?.[r.home]?.overPct ?? null,
          underPct: ou?.[r.home]?.underPct ?? null,
          totalPlusMinus: ou?.[r.home]?.totalPlusMinus ?? null
        }
      };

      games.push(r);
    }

    const completed = games.filter(
      (x) =>
        x.status !== "MISSING_ODDS" &&
        x.status !== "SKIPPED" &&
        Number.isFinite(x.homeScore) &&
        Number.isFinite(x.awayScore)
    );

    // All games with scores — includes SKIPPED (early season, MIN_GP not met)
    // Used for H2H learning even when projections aren't available yet
    const allScored = games.filter(
      (x) =>
        x.status !== "MISSING_ODDS" &&
        Number.isFinite(x.homeScore) &&
        Number.isFinite(x.awayScore)
    );

    const inBurnIn = i > (days - BURN_IN_DAYS);

    // Collect graded games for next day's processing
    // completed → Kalman + self-tune (needs projections)
    // allScored → H2H (just needs team names + scores)
    for (const g of completed) {
      g._kalmanDate = date;
    }
    prevGraded = completed;
    prevAllScored = allScored;

    const run = {
      date,
      dateDisplay,
      burnIn: inBurnIn,
      weightsUsed: { ...baseW },
      weightsNext: { ...baseW },
      games,
      summaryText: ""
    };

    upsertRun(store, run);
    saveStore(store);

    // Save Kalman state every 10 dates — if backfill crashes, we don't lose everything
    if (kalmanState && i % 10 === 0) {
      saveKalmanState(kalmanState);
    }

    const counts = games.reduce((acc, x) => {
      const k = x.status || "OK";
      acc[k] = (acc[k] || 0) + 1;
      return acc;
    }, {});
    const burnInTag = inBurnIn ? " [BURN-IN]" : "";
    const b2bCount = Object.keys(b2bNotes).length;
    console.log(
      `Backfilled ${dateDisplay}${burnInTag}: games=${games.length}, completed=${completed.length} statuses=${JSON.stringify(counts)}` +
      (b2bCount ? ` b2b=${b2bCount}` : "")
    );
  }

  // Save final Kalman state
  if (kalmanState) {
    pruneProcessedGames(kalmanState, 60);
    saveKalmanState(kalmanState);
    console.log("Kalman state saved.");
  }

  console.log("Backfill complete.");
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});